aiohttp/payload.py

```diff
@@ -197,20 +232,33 @@ def __init__(self, value, *args,
             if content_type is None:
                 content_type = 'text/plain; charset=%s' % encoding
 
-        super().__init__(
-            value.encode(encoding),
-            encoding=encoding, content_type=content_type, *args, **kwargs)
+        if encoding is None:
```

Comment:
Ok, this bit I completely do not get. It's literally impossible for encoding to be None, based on the logic. But the typing says otherwise and we can't exactly redefine the value with `mypy`. We can't also do the 

```
_encoding: str
_content_type: str
```

letting the assignment to come later, because tests on `3.5` complains. We could've do the `typing.cast` but not really sure if that's correct. Sounds a bit hacky to me.

@asvetlov any suggestions ?

Reply:
Alternatively, we can keep that `raise ValueError` and add some `noqa` there, but honestly that sounds bad as well to me.

Reply:
Oh, and for the sake of discussion, this is the alternative:

```
class StringPayload(BytesPayload):

    def __init__(self,
                 value: str,
                 *args: Any,
                 encoding: Optional[str]=None,
                 content_type: Optional[str]=None,
                 **kwargs: Any) -> None:
        _encoding, _content_type = self._parse_args(encoding, content_type)
        super().__init__(
            value.encode(_encoding),
            encoding=_encoding,
            content_type=_content_type,
            *args,
            **kwargs)

    @staticmethod
    def _parse_args(
        encoding: Optional[str]=None,
        content_type: Optional[str]=None,
    ) -> Tuple[str, str]:
        if encoding and content_type:
            return encoding, content_type
        elif encoding is None and content_type is None:
            return 'utf-8', 'text/plain; charset=utf-8'
        elif encoding is None and content_type:
            mimetype = parse_mimetype(content_type)
            return (
                mimetype.parameters.get('charset', 'utf-8'),
                content_type,
            )
        else:
            return (
                cast(str, encoding),
                'text/plain; charset=%s' % cast(str, encoding),
            )
```

Reply:
Revert back the change, drop `if encoding is None` check.
Use the following trick:
```
        if encoding is None:
            if content_type is None:
                 real_encoding = 'utf-8'
                 content_type = 'text/plain; charset=utf-8'
             else:
                 mimetype = parse_mimetype(content_type)
                 readl_encoding = mimetype.parameters.get('charset', 'utf-8')
         else:
            if content_type is None:
                content_type = 'text/plain; charset=%s' % encoding
            real_encoding = encoding

 super().__init__(
                value.encode(encoding),
                encoding=real_encoding,
                content_type=content_type,
                *args,
                **kwargs)
```

`real_encoding` variable has strict string type, not `Optional[str]`.

I believe mypy understands it pretty well.